src/processor/operator/copy/copy_node.cpp

```diff
@@ -81,145 +69,163 @@ std::pair<std::string, common::row_idx_t> CopyNode::getFilePathAndRowIdxInFile()
+        while (numAppendedTuples < numTuplesToAppend) {
+            numAppendedTuples += localNodeGroup->append(
+                resultSet, copyNodeInfo.dataColumnPoses, numTuplesToAppend - numAppendedTuples);
+            if (localNodeGroup->getNumNodes() == StorageConstants::NODE_GROUP_SIZE) {
+                // Current node group is full, flush it and reset it to empty.
+                auto nodeGroupIdx = sharedState->getNextNodeGroupIdx();
+                localNodeGroup->setNodeGroupIdx(nodeGroupIdx);
+                appendNodeGroupToTableAndPopulateIndex(sharedState->table, localNodeGroup.get(),
+                    sharedState->pkIndex.get(), sharedState->pkColumnID);
+            }
+            if (numAppendedTuples < numTuplesToAppend) {
+                auto slicedChunk = sliceDataVectorsInDataChunk(*dataChunkToCopy,
+                    copyNodeInfo.dataColumnPoses, (int64_t)numAppendedTuples,
+                    (int64_t)(numTuplesToAppend - numAppendedTuples));
+                resultSet->dataChunks[0] = slicedChunk;
+            }
         }
     }
-    return numValues;
+    // Append left data in the local node group to the shared one.
+    if (localNodeGroup->getNumNodes() > 0) {
+        sharedState->appendLocalNodeGroup(std::move(localNodeGroup));
+    }
 }
 
-template<>
-uint64_t CopyNode::appendToPKIndex<ku_string_t, InMemOverflowFile*>(InMemColumnChunk* chunk,
-    offset_t startOffset, uint64_t numValues, InMemOverflowFile* overflowFile) {
-    for (auto i = 0u; i < numValues; i++) {
-        auto offset = i + startOffset;
-        auto value = chunk->getValue<ku_string_t>(i);
-        auto key = overflowFile->readString(&value);
-        if (!sharedState->pkIndex->append(key.c_str(), offset)) {
-            return i;
-        }
-    }
-    return numValues;
+std::shared_ptr<DataChunk> CopyNode::sliceDataVectorsInDataChunk(const DataChunk& dataChunkToSlice,
+    const std::vector<DataPos>& dataColumnPoses, int64_t offset, int64_t length) {
+    auto slicedChunk =
+        std::make_shared<DataChunk>(dataChunkToSlice.getNumValueVectors(), dataChunkToSlice.state);
+    for (auto& dataPos : dataColumnPoses) {
+        slicedChunk->valueVectors[dataPos.valueVectorPos] =
+            std::make_shared<ValueVector>(LogicalTypeID::ARROW_COLUMN);
+    }
+    for (auto& dataColumnPose : dataColumnPoses) {
+        assert(dataColumnPose.dataChunkPos == 0);
+        auto vectorPos = dataColumnPose.valueVectorPos;
+        ArrowColumnVector::slice(dataChunkToSlice.valueVectors[vectorPos].get(),
+            slicedChunk->valueVectors[vectorPos].get(), offset, length);
+    }
+    return slicedChunk;
+}
+
+void CopyNode::appendNodeGroupToTableAndPopulateIndex(NodeTable* table, NodeGroup* nodeGroup,
+    PrimaryKeyIndexBuilder* pkIndex, column_id_t pkColumnID) {
+    auto numNodes = nodeGroup->getNumNodes();
+    auto startOffset = nodeGroup->getNodeGroupIdx() << StorageConstants::NODE_GROUP_SIZE_LOG2;
```

Comment:
Maybe refactor this to a function `getStartOffsetForGroup()` ?

Reply:
See `StorageUtils::getStartOffsetForNodeGroup`